distributed/diagnostics/plugin.py

```diff
@@ -282,34 +278,73 @@ def __init__(self, packages, pip_options=None, restart=False):
         self.pip_options = pip_options or []
+        self.id = f"pip-install-{uuid.uuid4()}"
 
     async def setup(self, worker):
-        from distributed.lock import Lock
-
-        async with Lock(socket.gethostname()):  # don't clobber one installation
-            logger.info("Pip installing the following packages: %s", self.packages)
-            proc = subprocess.Popen(
-                [sys.executable, "-m", "pip", "install"]
-                + self.pip_options
-                + self.packages,
-                stdout=subprocess.PIPE,
-                stderr=subprocess.PIPE,
-            )
-            stdout, stderr = proc.communicate()
-            returncode = proc.wait()
-
-            if returncode:
-                logger.error("Pip install failed with '%s'", stderr.decode().strip())
-                return
-
-            if self.restart and worker.nanny:
-                lines = stdout.strip().split(b"\n")
-                if not all(
-                    line.startswith(b"Requirement already satisfied") for line in lines
-                ):
-                    worker.loop.add_callback(
-                        worker.close_gracefully, restart=True
-                    )  # restart
+        from distributed.semaphore import Semaphore
+
+        async with (
+            await Semaphore(max_leases=1, name=socket.gethostname(), register=True)
+        ):
+            if not await self._is_installed(worker):
+                logger.info("Pip installing the following packages: %s", self.packages)
+                await self._set_installed(worker)
+                proc = subprocess.Popen(
+                    [sys.executable, "-m", "pip", "install"]
+                    + self.pip_options
+                    + self.packages,
+                    stdout=subprocess.PIPE,
+                    stderr=subprocess.PIPE,
+                )
+                stdout, stderr = proc.communicate()
+                returncode = proc.wait()
+                if returncode != 0:
+                    msg = f"Pip install failed with '{stderr.decode().strip()}'"
+                    logger.error(msg)
+                    raise RuntimeError(msg)
+            else:
+                logger.info(
+                    "The following packages have already been installed: %s",
+                    self.packages,
+                )
+
+            if self.restart and worker.nanny and not await self._is_restarted(worker):
+                logger.info("Restarting worker to refresh environment")
+                await self._set_restarted(worker)
+                worker.loop.add_callback(worker.close_gracefully, restart=True)
+
+    async def _is_installed(self, worker):
+        return await worker.client.get_metadata(
+            self._compose_installed_key(), default=False
+        )
+
+    async def _set_installed(self, worker):
+        await worker.client.set_metadata(
+            self._compose_installed_key(),
+            True,
+        )
```

Comment:
Setting metadata needs another lock. The above lock is merely per host. This needs a global lock